### ISSUE TITLE
<address>.call{value: value}(new bytes(0)) causes ERROR call may fail or behave unexpectedly due to empty code

### ISSUE THREAD

**Issue by vmaark** [matter-labs/foundry-zksync#793]
### Component

Forge

### Have you ensured that all of these are up to date?

- [X] Foundry
- [X] Foundryup

### What version of Foundry are you on?

forge 0.0.2 (68fcb7d 2024-12-16T00:45:46.210537000Z)

### What command(s) is the bug in?

forge test

### Operating System

macOS (Apple Silicon)

### Describe the bug

Sending eth with .call causes the following log in my forge test output: 
`2024-12-18T22:55:36.093440Z ERROR foundry_zksync_core::vm::tracers::cheatcode: call may fail or behave unexpectedly due to empty code target=0x0000000000000000000000000000000000989682 calldata=""`

Although it displays it as an `ERROR`, the execution continues. Due to having a lot of tests and fuzzing, this really pollutes my console.

Proposed solution:
When calldata is empty, due not assume it's a function call, don't check the address for code.

**Comment by vmaark** [comment]
Also I suspect that this is making somehow my tests running extremely slow (30 minutes on github actions, instead of like 5 mins), not sure if there's some retry mechanism under the hood, or it's just the fuzzing, but I have 100s of lines of the same  error being reported (and ignored, execution continues).

**Comment by Jrigada** [comment]
Hey @vmaark can you share an example of a test that is printing the error? A simple <address>.call{value: value}("") is not printing it out for me. 
```
contract A {
    receive() external payable {}
}

contract AddressCallTest is Test {
    A internal a;
    function setUp() public {
        a = new A();
    }
    function testAddressCall() public {
        (bool success, ) = address(a).call("");
        assertTrue(success);
    }

    function testAddressCallValue() public {
        (bool success, ) = address(a).call{value: 1 ether}("");
        assertTrue(success);
        assertEq(address(a).balance, 1 ether);
    }
}
```

**Comment by vmaark** [comment]
> contract A {
>     receive() external payable {}
> }
> 
> contract AddressCallTest is Test {
>     A internal a;
>     function setUp() public {
>         a = new A();
>     }
>     function testAddressCall() public {
>         (bool success, ) = address(a).call("");
>         assertTrue(success);
>     }
> 
>     function testAddressCallValue() public {
>         (bool success, ) = address(a).call{value: 1 ether}("");
>         assertTrue(success);
>         assertEq(address(a).balance, 1 ether);
>     }
> }

in this example you're sending to a contract which works fine. It happens when sending eth to an EOA using .call - sorry if that wasn't clear from the original description.

**Comment by Jrigada** [comment]
Hey @vmaark we just merged a fix for this exact case. It should spare you from the error you were getting. Thanks for raising it.

**Comment by vmaark** [comment]
hey @Jrigada, thanks for the quick fix, sounds great! Updated zksync-forge, I assume it is not yet included in the following version:
```
✗ forge --version
forge 0.0.2 (04e701d 2025-01-09T00:34:37.572492000Z)
```

**Comment by Jrigada** [comment]
Yes, is not present yet in the release. To use it right now you would have to build directly from main. Probably by today or early next week it will be on the new release!

**Comment by Karrq** [comment]
Hey @vmaark could you let us know if this fixed your issue?
If so, please feel free to close this issue, thank you

**Comment by vmaark** [comment]
> Hey [@vmaark](https://github.com/vmaark) could you let us know if this fixed your issue? If so, please feel free to close this issue, thank you

thanks for the ping!
```bash
(base) ➜  git:(zksync) ✗ forge --version
forge 0.0.2 (5c46959 2025-01-21T00:24:28.347259000Z)
```

getting the same error

**Comment by Karrq** [comment]
@Jrigada this is the related PR, right? https://github.com/matter-labs/foundry-zksync/pull/804 
Should have been included then

**Comment by vmaark** [comment]
> [@Jrigada](https://github.com/Jrigada) this is the related PR, right? [#804](https://github.com/matter-labs/foundry-zksync/pull/804) Should have been included then

sorry, do you mean it was supposed to and it didn't, or some other issue is happening?

**Comment by Karrq** [comment]
I meant that if the fix was done thru #804 then you shouldn't be seeing the issue anymore, so I was double checking with @Jrigada that the PR in question is actually the fix, and if it is then we should investigate why you are still observing the error :) 

So yeah, it was supposed to fix it (from what I can tell) but didn't

**Comment by Jrigada** [comment]
Hey jumping in here!
The error should not appear when using `call()` to an address with a certain value. We have a couple of tests [here](https://github.com/matter-labs/foundry-zksync/blob/5c4695922dde0cbce1cf06aefab2ea5bd8d4c9dc/crates/forge/tests/cli/zk_cmd.rs#L126). I also made sure `makeAddr` is not persisting the account behind the scenes or something along those lines.  
Could you provide the snippet of code you are testing to reproduce it? We might have missed something. From what I understood the error was happening when sending funds to an arbitrary address (not a contract).

**Comment by Jrigada** [comment]
Okay I tried with the address from your logs and it did throw the error! I will continue debugging and update as soon as possible.

**Comment by vmaark** [comment]
> ~Okay I tried with the address from your logs and it did throw the error! I will continue debugging and update as soon as possible~ I was trying a call with calldata, sorry for the false alarm. If you have a reproducible example that would be great.

great, I was unsure for a second, because I couldn't reproduce it - what I saw when I wrote my previous reply was valid, there was non-empty calldata, so the issue can bel closed 🙏 

Sidenote - `call may fail or behave unexpectedly due to empty code - please ensure the current calldata is mocked` - shouldn't it say "please ensure the current **target** is mocked"?

**Comment by Karrq** [comment]
> shouldn't it say "please ensure the current target is mocked"?

That specific bit is because the target address is _already_ being mocked, it's just the input calldata that differs from what was specified in the mock. 
If the address wasn't being mocked it wouldn't show the [second section of the message](https://github.com/matter-labs/foundry-zksync/blob/dafc7a444bab46d36d44090493f398cc3f84076e/crates/zksync/core/src/vm/tracers/cheatcode.rs#L310-L313) ;) 

Glad to know the issue is fixed <3
